### PR TITLE
Reduce code duplication

### LINKED ISSUE #56: Reduce code duplication
A lot of the code in addon and consumer is duplicated.

Proposed solution: create a utility package that both the addon and consumer can use and anything that is duplicated should be added to the package and removed from the base code for the service.

### PR DESCRIPTION
**Pull request by purple-emily**
This will close #56.

I'd like some help with this when anyones available, but I don't mind doing what I can for now. Ideally @iPromKnight you'll add me on Discord and help me quickly work out if I have the correct dev environment set up for JS 😂

### CODE REVIEW

**Comment by sleeyax** [comment]
I could also help with this. 

To orchestrate the packages we could set up a simple [NPM workspace](https://docs.npmjs.com/cli/v7/using-npm/workspaces) in the `node` folder. The structure then becomes as follows:

```
├── addon
│   ├── package.json
├── addon-jackett
│   ├── package.json
├── consumer
│   ├── package.json
└── lib
    └── package.json
├── package.json
```

When everything is setup correctly you can then `npm install lib` in `addon` and `consumer`, which will symlink the shared `lib` package and make it available for import:

```js
import { example } from 'lib';
// ...
```

**Comment by sleeyax** [comment]
Another upside to the workspace approach is you can share `devDependendies` between all packages by defining them once in the parent `package.json` file (`node/package.json` in this case).

File: src/node/addon/package.json [hunk 1]
```diff
@@ -19,6 +19,7 @@
     "debrid-link-api": "^1.0.1",
     "express-rate-limit": "^6.7.0",
     "ip": "^1.1.8",
+    "knightcrawler-utility": "file:../utility",
```
Comment (sleeyax) [inline src/node/addon/package.json:22]: Prefixing with the project name isn't really required as this utility package is never exposed (i.e. installable) to the outside world.

File: src/node/utility/package.json [hunk 2]
```diff
@@ -0,0 +1,9 @@
+{
+  "name": "knightcrawler-utility",
```
Comment (sleeyax) [inline src/node/utility/package.json:2]: Alternatively you can add a [scope](https://docs.npmjs.com/cli/v7/using-npm/scope) like `@knightcrawler/utility`, the same style applies to the other packages in the format `@project/package`. Though it's not really required as just `utilities` works too.

**Comment by purple-emily** [comment]
@sleeyax is lib a common convention in these situations for JS? I was trying to decide what the best name would be

**Comment by purple-emily** [comment]
@sleeyax you're more than welcome to fork and commit some changes and make a pull request 😆

**Comment by sleeyax** [comment]
Yep, was just about to do so @purple-emily you read my mind :)

**Comment by iPromKnight** [comment]
Yep. Workspaces is what we should use.
I was gonna add Nx though - so we can cache build artifacts for individual package builds within git and share between us
No Nx cloud though - all disconnected

https://nx.dev/getting-started/intro
https://nx.dev/nx-api/esbuild

**Comment by iPromKnight** [comment]
> This will close #56.
> 
> I'd like some help with this when anyones available, but I don't mind doing what I can for now. Ideally @iPromKnight you'll add me on Discord and help me quickly work out if I have the correct dev environment set up for JS 😂

Ahh I hardly use it these days - you should see my unread message count hahaha
Sure can do - but it sounds like @sleeyax might be a better fit

I'm a .net dev really 23 years of it - this is the first JS project i've worked on in the past 4 years  - although it seems not much has changed just tooling has gotten better ^^

**Comment by purple-emily** [comment]
> Yep. Workspaces is what we should use, rather than sub npm packages or github packages
> 
> I was gonna add Nx though - so we can cache build artifacts for individual package builds within git and share between us
> 
> No Nx cloud though - all disconnected
> 
> 
> 
> https://nx.dev/getting-started/intro
> 
> https://nx.dev/nx-api/esbuild
> 
> 
> 
> It handles caching of build and test artifacts, less duplication of config files etc
> 
> Plus i'd like to start rewriting some of our ES modules as typescript to get some damn type checking added for dev lol
> 
> All this dynamic typing makes me feel like i need 5 showers a day 👅 

Don't let me stop you 😂 I don't know what @sleeyax has done already

**Comment by sleeyax** [comment]
Oh okay, I've used `nx` before so can add the base setup for that too. Not sure if a shared cache is warranted (as in *really* needed) at the current stage of this project though, perhaps this should be a new issue for improvement? 

>  Plus i'd like to start rewriting some of our ES modules as typescript to get some damn type checking added for dev lol

Perfect. It's actually insane this project isn't using typescript already haha.

**Comment by iPromKnight** [comment]
Really is 

> Oh okay, I've used `nx` before so can add the base setup for that too. Not sure if a shared cache is warranted (as in _really_ needed) at the current stage of this project though, perhaps this should be a new issue for improvement?
> 
> > Plus i'd like to start rewriting some of our ES modules as typescript to get some damn type checking added for dev lol
> 
> Perfect. It's actually insane this project isn't using typescript already haha.

Haha I hear ya - so frustrating
It was all pure js when we originally took the old code too, had the pain of changing all imports, and introducing esbuild so we could transpile

**Comment by sleeyax** [comment]
Yeah I was surprised to see `tsx` being used in a non-typescript project too :laughing:

**Comment by iPromKnight** [comment]
> Yeah I was surprised to see `tsx` being used in a non-typescript project too 😆

Haha - I only added recently - but that was part of the plan to move us forward. -plus we can still use watch, and pipe it into pino now i've added that too for pretty logs during dev

**Comment by iPromKnight** [comment]
Please consider moving moch's to a separate lib too, That way our only upstream (torrentio) commits to mochs for the time being will be to that lib - nothing else should hopefully have to change

And once we have the moch's rewritten - we'll have to support those too

Who knows - maybe Torrentio's author will end up pulling some of the improvements from us lol

**Comment by sleeyax** [comment]
Ouch setting up `nx` is a little more complex than I remember, it pretty much requires a `project.json` per project with the `esbuild` executor configured which in turn requires a `tsconfig.json` file to be present per project which is iffy since we're not using typescript yet :sweat_smile: 

I'll go ahead and create a PR for the NPM workspace restructure first so @purple-emily can continue moving shared code to the common lib. The `nx` setup can then be continued in parallel.

**Comment by iPromKnight** [comment]
> Ouch setting up `nx` is a little more complex than I remember, it pretty much requires a `project.json` per project with the `esbuild` executor configured which in turn requires a `tsconfig.json` file to be present per project which is iffy since we're not using typescript yet 😅
> 
> I'll go ahead and create a PR for the NPM workspace restructure first so @purple-emily can continue moving shared code to the common lib. The `nx` setup can then be continued in parallel.
> 
> See #60

Might be worth holding off on moving stuff to the shared codebase right now - The addon is 85% typescript in my pr draft now in #69 
Wouldn't mind you having a look over ^^ ta

**Pull request closed without merging**